src/test/kotlin/cacophony/automata/minimalization/DFAMinimalizationTest.kt

```diff
@@ -0,0 +1,177 @@
+                    18 via 'p' to 19,
+                    19 via 'h' to 12,
+                    // unreachable state
+                    20 via 'x' to 7,
+                ),
+            )
+        val minimized = dfa.minimalize()
+        assertTrue(areEquivalent(dfa, minimized))
+        assertEquals(16, minimized.getAllStates().size)
+    }
+
+    private fun <E> checkThatMinimalizedDFAIsEquivalent(dfa: DFA<E>) {
+        val minDfa = dfa.minimalize()
+        val helper = createDFAEquivalenceHelper(dfa, minDfa)
+
+        for (state in minDfa.getAllStates()) {
+            for (oldState in state.originalStates) {
+                assert(helper.areEquivalent(oldState, state))
+            }
+        }
+
+        assert(helper.areEquivalent(dfa.getStartingState(), minDfa.getStartingState()))
+    }
+
+    private fun <E> bruteDFAStatesEquivalenceClasses(dfa: DFA<E>): Set<Set<E>> {
+        val helper = createDFAEquivalenceHelper(dfa, dfa)
+        val states = dfa.getAllStates()
+
+        val equivalenceClassesMap: MutableMap<E, MutableSet<E>> = mutableMapOf()
+        states.forEach {
+            for (s in states) {
+                if (helper.areEquivalent(s, it)) {
+                    equivalenceClassesMap.getOrPut(s) { mutableSetOf() }.add(it)
+                    return@forEach
+                }
+            }
+            throw Exception("State is not equivalent to itself")
+        }
+        return equivalenceClassesMap.values.toSet()
+    }
+
+    private fun <E> checkThatMinimalizedDFAIsMinimal(dfa: DFA<E>) {
+        val equivalenceClassesExpected = bruteDFAStatesEquivalenceClasses(dfa.withAliveReachableStates())
+        val equivalenceClassesActual =
+            dfa
+                .minimalize()
+                .getAllStates()
+                .map { it.originalStates.toSet() }
+                .toMutableSet()
+
+        assertEquals(equivalenceClassesExpected, equivalenceClassesActual)
+    }
+
+    private fun <E> checkRandomDFA(dfa: DFA<E>) {
+        if ((dfa.getAliveStates() intersect dfa.getReachableStates()).isEmpty()) {
+            try {
+                dfa.minimalize()
+                TODO("Fix tests once DFA interface allows empty language")
+            } catch (e: NotImplementedError) {
+                return
+            }
+        }
+
+        checkThatMinimalizedDFAIsEquivalent(dfa)
+        checkThatMinimalizedDFAIsMinimal(dfa)
+    }
+
+    private fun generateRandomDFA(
+        n: Int,
+        random: Random,
+    ): DFA<Int> {
+        val density = 0.2
+        val symbols = "abc"
+        val states = 1..n
+        return createDFA(
+            states.random(random),
+            states.filter { random.nextDouble() < 0.2 }.toSet(),
+            (0..<ceil(density * n * symbols.length).toInt()).associate {
+                states.random(random) via symbols.random(random) to states.random(random)
+            },
+        )
+    }
+
+    @Test
+    fun `Random DFAs`() {
```

Comment:
I'm not sure about this, imo tests shouldn't be non-deterministic. If implementation is wrong it can randomly fail on build not related to this. We can leave it now and discuss it on Tuesday.

Reply:
Tests are deterministic as they use fixed seeds. 

java.util.Random guarantees this https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/Random.html

Kotlin's random uses slightly weaker wording: "Note: Future versions of Kotlin may change the algorithm of this seeded number generator so that it will return a sequence of values different from the current one for a given seed." I personally do not think this is a problem, but I can change the random to java one if needed.